### DISCUSSION TITLE
Migration to runtimeConfig

### DISCUSSION THREAD

**Discussion by pi0** [nuxt/nuxt#7950] [General]
[originally asked by @hecktarzuli] 

> I'm curious if there is a plan or general 'movement' to migrate from nuxt.config.js options vs runtime configs, or a general practice to combine both for community modules.. I would think modules would either look at both, or combine them .. leading the way in the future for realtime configs to modify either.

**Comment by pi0** [comment]
Actually there are no general practices to support runtimeConfig for **modules** as it is mostly a runtime functionality so should be directly supported in **plugins** added by modules and varies by use case.

Here is an implementation example: https://github.com/nuxt/http/blob/dev/lib/plugin.js#L186 

Basically inside plugin's export we should do:

```js
// plugin.js
export default (ctx, inject) => {
  const runtimeConfig = ctx.$config && ctx.$config.moduleKey || {}
  // We can modify plugin behavior at runtime based on runtime-provided config 
}
```

Module function always get fresh config also for production. Only if they need to modify plugin behavior at runtime based on config, need runtimeConfig support. In case of a module-computed option required by plugin (like when modules needs to parse a file and provide parsed file to runtime plugin), modules can directly modify `nuxt.options.publicRuntimeConfig`. Example:

```js
// module.js
export default function () { 
  const { nuxt } = this
  const { publicRuntimeConfig } = nuxt.options

  publicRuntimeConfig.myModule = publicRuntimeConfig.myModule || {}
  publicRuntimeConfig.myModule.foobar = '<a computed prop>'
}
```

Note: Modification by module, can happen anytime during module lifecycle. So modules can even change it's value multiple times like using a timer.


PS: Maybe we can add some more docs on this area :)

**Reply by hecktarzuli** [reply]
When I say `community modules` I was talking generally about the whole libs, not literally Nuxt Modules (sorry for the confusion).  My general feeling is since we provide 2 ways to setup a config in nuxt, all new libs should probably support both out of the box.

**Reply by pi0** [reply]
Technically any community nuxt module is also a nuxt module :D They can optionally/partially support runtimeConfig but shouldn't necessarily do or always can do (like sometimes a build-time configuration needs to be updated)

**Reply by pi0** [reply]
Also as mentioned `nuxt.config` is runtime too! It will be read everytime on production and can be updated at runtime using `nuxt.options` inside modules. RuntimeConfig lives into another context which is webpack runtime (either SSR or Browser).

**Reply by hecktarzuli** [reply]
Sure.  My point is, any new community nuxt module that the internal team creates should probably try to support both when it makes sense.  That's all :)

**Reply by pi0** [reply]
Hmm we may document this line then:

>   const runtimeConfig = ctx.$config && ctx.$config.moduleKey || {}

:)

**Reply by hecktarzuli** [reply]
Great info, thanks!

**Reply by pi0** [reply]
@hecktarzuli Just created nuxt/nuxtjs.org#532 for docs. We may also add it to [module template](https://github.com/nuxt-community/module-template/blob/template/templates/plugin.js#L5)
